packages/client/src/game-core/utils/common-types.ts

```diff
@@ -0,0 +1,44 @@
+export type Nullable<T> = T | null | undefined
+export type Primitive = boolean | number | bigint | string
+export type NonEmptyArr<T> = [T, ...T[]]
+
+export type Indexed<T = unknown> = {
+  [k in string]: T
+}
+
+export type NumIndexed<T = unknown> = {
+  [k in number]: T
+}
+
+export type Readonly<T extends Indexed | NumIndexed> = {
+  readonly [P in keyof T]: T[P]
+}
+
+export type DeepReadonly<T> = {
+  readonly [P in keyof T]: T[P] extends (infer U)[]
+    ? DeepReadonly<U>[]
+    : T[P] extends object
+    ? DeepReadonly<T[P]>
```

Comment:
Отступы нарушены. Или это Prettier так отформатировал?
```suggestion
      ? DeepReadonly<T[P]>
```